kive/container/test_dataset_comparison.py

```diff
@@ -0,0 +1,596 @@
+    @mock.patch("container.runutils._compare_optional_inputs")
+    @mock.patch("container.runutils._compare_monovalent_args")
+    def test_comparison_selection(self, compare_mono, compare_optional,
+                                  compare_dir):
+        "Verify that `compare_rerun_datasets` chooses the right comparison function"
+
+        def reset_mocks():
+            compare_mono.reset_mock()
+            compare_optional.reset_mock()
+            compare_dir.reset_mock()
+
+        def make_mock_run(argtype):
+            arg = mock.Mock()
+            arg.argtype = argtype
+            run = mock.Mock()
+            run.app = mock.Mock()
+            run.app._argument = arg
+            run.app.arguments.all = mock.Mock(return_value=[arg])
+            return run
+
+        cases = [
+            (ContainerArgumentType.FIXED_INPUT, [compare_mono]),
+            (ContainerArgumentType.FIXED_OUTPUT, [compare_mono]),
+            (ContainerArgumentType.OPTIONAL_INPUT, [compare_optional]),
+            (ContainerArgumentType.OPTIONAL_MULTIPLE_INPUT, [compare_optional
+                                                             ]),
+            (ContainerArgumentType.FIXED_DIRECTORY_OUTPUT, [compare_dir]),
+        ]
+
+        for argtype, expect_called in cases:
+            run = make_mock_run(argtype)
+            runutils.compare_rerun_datasets(run, run)
+            for comparison_mock in [
+                    compare_mono, compare_optional, compare_dir
+            ]:
+                if comparison_mock in expect_called:
+                    comparison_mock.assert_called_with(run.app._argument, run, run)
+                else:
+                    comparison_mock.assert_not_called()
+            reset_mocks()
+
+
+class BaseDatasetComparisonTestCase(TestCase):
+    """Common setup for testing dataset comparison functions.
+
+    Test cases that inherit from this class will have:
+
+    - A ContainerApp (with associated Container and ContainerFamily, which 
+      we don't care about but create to satisfy foreign key requirements.)
+    - Datasets containgin 'a', 'b', and ''.
+
+    Test cases that inherit from this class can add their own arguments to
+    the app (they'll be dropped when the tests finish) and define their own
+    ContainerRun and ContainerDataset instances.
+    """
+
+    # Pre-computed MD5 values for 'a', 'b', and ''.
+    MD5_A = "60b725f10c9c85c70d97880dfe8191b3"
+    MD5_B = "3b5d5c3712955042212316173ccf37be"
+    MD5_EMPTY = "68b329da9893e34099c7d8ad5cb9c940"
+
+    @classmethod
+    def setUpTestData(cls) -> None:
+        super().setUpTestData()
+        cls._kive_user = metadata.models.kive_user()
+
+        cls.containerfamily = ContainerFamily.objects.create(
+            name="Dataset Comparison Test Family",
+            description="",
+            git="",
+            user=cls._kive_user,
+        )
+        cls.containerfamily.save()
```

Comment:
I'm pretty sure you don't need to call `save()` after `create(...)`.

Reply:
The test suite agrees. :+1: